### ISSUE TITLE
Can't codesign aarch64-ios binaries on Intel mac

### ISSUE THREAD

**Issue by jmrico01** [ziglang/zig#9565]
I am able to build a full iOS app with Zig after #9532 - yay! And https://github.com/kubkon/zig-ios-example is a great simplified example of how to do that and run it in a simulator. However, something is preventing me from code-signing the app, which is required to run it on a real iOS device. I am using `zig-macos-x86_64-0.9.0-dev.796+fcf2ce0ff`, and running the following on @kubkon's example repo:

- `zig build -Dtarget=aarch64-ios`
- `codesign -s "<my-adhoc-certificate>" zig-out/bin/MadeWithZig.app`

The code signing fails with
`zig-out/bin/MadeWithZig.app: is already signed`
So I tried using the `-f`/`--force` flag:
- `codesign -f -s "<my-adhoc-certificate>" zig-out/bin/MadeWithZig.app`
and it spilled the beans:
```
zig-out/bin/MadeWithZig.app: replacing existing signature
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/codesign_allocate: fatal error: file not in an order that can be processed (code signature data out of place): /Users/legionjr/Documents/zig-ios-example/zig-out/bin/MadeWithZig.app/app
zig-out/bin/MadeWithZig.app: the codesign_allocate helper tool cannot be found or used
```
To inspect this existing signature, I ran
- `codesign -dv zig-out/bin/MadeWithZig.app`
which prints
```
Executable=/Users/legionjr/Documents/zig-ios-example/zig-out/bin/MadeWithZig.app/app
Identifier=app
Format=app bundle with Mach-O thin (arm64)
CodeDirectory v=20400 size=252 flags=0x2(adhoc) hashes=5+0 location=embedded
Signature=adhoc
Info.plist=not bound
TeamIdentifier=not set
Sealed Resources=none
Internal requirements=none
```

Poking around the Zig source, I see there is some logic now that throws in a code signature during MachO linking. Maybe this is something that was put in for macOS, but should be done differently for iOS builds? Not familiar with the expected format of this code signature, but `codesign` definitely doesn't like it...

**Comment by kubkon** [comment]
Thanks for the report @jmrico01! The presence of adhoc code signature shouldn't be a problem - the actual issue is a mismatch in expectations between what Zig produces as a binary versus what codesign_allocate tool actually expects and treats as valid. For the record, I have it on my agenda to rewrite notorisation tool (codesign) in Zig and make it platform independent (so that anyone can create a full sig for iOS from anywhere) but in the meantime, I'll investigate what codesign_allocate sees as invalid. It might ordering of load commands and corresponding Linkedit sections, or existing CodeSignature size not being properly aligned. I'll keep you posted either way!

**Comment by kubkon** [comment]
Oh wait, I think you're pointing `codesign` to the wrong thing. You're pointing it to the app folder `MadeWithZig.app` whereas you should point it to the actual binary so `MadeWithZig.app/app`:

```
> codesign -s "<cert>" zig-out/bin/MadeWithZig.app/app
```

For the record, I just tried rewriting the adhoc code signature already present in `app` with Apple's `codesign` and everything worked fine for me:

```
> codesign -f -s - zig-out/bin/MadeWithZig.app/app
zig-out/bin/MadeWithZig.app/app: replacing existing signature
```

**Comment by jmrico01** [comment]
phew, that was almost embarrassing... that's not it unfortunately - I see the same output in both cases, whether I point `codesign`to the app folder or the actual executable. I guess codesign looks for the `app` binary on my behalf.

It's weird that yours is working, though. Maybe we have different versions of `codesign`? I'm on an old 2013 macbook air running macOS Big Sur (11.4). My Xcode version is 12.5.1 (12E507) (I'm assuming `codesign` is bundled with Xcode? can't get `codesign` itself to print a version number). Still compiling with `zig-macos-x86_64-0.9.0-dev.796+fcf2ce0ff`.

**Comment by kubkon** [comment]
> phew, that was almost embarrassing... that's not it unfortunately - I see the same output in both cases, whether I point `codesign`to the app folder or the actual executable. I guess codesign looks for the `app` binary on my behalf.

Ah, that's good to know actually, thanks!

> It's weird that yours is working, though. Maybe we have different versions of `codesign`? I'm on an old 2013 macbook air running macOS Big Sur (11.4). My Xcode version is 12.5.1 (12E507) (I'm assuming `codesign` is bundled with Xcode? can't get `codesign` itself to print a version number). Still compiling with `zig-macos-x86_64-0.9.0-dev.796+fcf2ce0ff`.

Hmm, I've checked on macOS 12 and Xcode 13, so lemme do another run on macOS 11.4 and Xcode 12.5 and see if it fails for me too.

**Comment by kubkon** [comment]
Good news: I can repro this on my Intel MBP. Interesting fact: no problem on the M1 however. I will investigate why that is.

**Comment by kubkon** [comment]
> Good news: I can repro this on my Intel MBP. Interesting fact: no problem on the M1 however. I will investigate why that is.

I still don't fully understand why `codesign` behaves differently on M1 vs Intel Macs (something tells me it treats an iOS binary as if it was meant to run on the Mac and hence accepts the adhoc code signature we generate with Zig which is clearly rejected by `codesign` on the Intel Mac). Anyhow, I believe the best way forward for the foreseeable future is to generate and embed adhoc code signature only for binaries which target `aarch64-macos`. For any other platform, we simply won't do that  referring the user to `codesign` utility which as far as I understand is required if you want to install your app/bundle on an actual iPhone, etc.
